Various cmake cleanups

rschu1ze opened the pull request:
More cmake simplification and cleanup while teaching me cmake in preparation for another task.

See the individual commits for details.

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

rschu1ze committed:
Throw option WEVERYTHING out

WEVERYTHING enables on Clang literally every warning. People on the
internet are divided if this is a good thing or not but ClickHouse
compiles with -Weverything + some exceptions for noisy warnings since at
least a year.

I tried to build with WEVERYTHING = OFF and the build was badly broken.
It seems nobody actually turns WEVERYTHING off. Actually, why would one
if the CI builds (configured with WEVERYTHING = ON) potentially generate
errors not generated in local development.

To simplify the build scripts and to remove the need to maintain two
sets of compiler warnings, I made WEVERYTHING the default and threw
WEVERYTHING = OFF out.

rschu1ze committed:
Fixing some warnings

rschu1ze committed:
Cosmetics

rschu1ze committed:
Fix externally set compiler launchers

The compiler launcher (ccache, distcc) can be set externally via
-DCMAKE_CXX_COMPILER_LAUNCHER=<tool>. We previously silently ignored
this setting and continued without any launcher (e.g. ccache). Changed
this to now respect the externally specified launcher.

rschu1ze committed:
Use existing variable COMPILER_CLANG to identify Clang

rschu1ze committed:
Made the ccache integration more straightforward

The new logic is simpler and should do the same as before.

rschu1ze committed:
Remove weird check for not-Clang

The check activated ccache unconditionally for all non-Clang compilers
(= GCC) while allowing ancient ccache versions for these. Perhaps there
was a reason for that in the past but it's simpler to only require a
minimum ccache version.

To simplify further, also require at least ccache 3.3 (released in 2016)
instead of 3.2.1 (released in 2014).

rschu1ze committed:
GLOB with CONFIGURE_DEPENDS

Globbing generally misses to pick up files which were added/deleted
after CMake's configure. This is a nuissance but can be alleviated using
CONFIGURE_DEPENDS (available since CMake 3.12) which adds a check for
new/deleted files before each compile and - if necessary - restarts the
configuration. On my system, the check takes < 0.1 sec.

(Side note: CONFIGURE_DEPENDS is not guaranteed to work accross all
generators, but at least it works for Ninja which everyone @CH seems to
use.)

azat commented on cmake/ccache.cmake:
```diff
@@ -1,56 +1,51 @@
+set (CCACHE_MINIMUM_VERSION 3.3)
+
+if (CCACHE_VERSION VERSION_LESS_EQUAL ${CCACHE_MINIMUM_VERSION})
+    message(FATAL_ERROR "Using ccache: no (found ${CCACHE_EXECUTABLE} (version ${CCACHE_VERSION}), the minimum required version is ${CCACHE_MINIMUM_VERSION}")
```
```suggestion
    message(${RECONFIGURE_MESSAGE_LEVEL} "Using ccache: no (found ${CCACHE_EXECUTABLE} (version ${CCACHE_VERSION}), the minimum required version is ${CCACHE_MINIMUM_VERSION}")
    return()
```

rschu1ze replied:
I wasn't exactly sure either but your proposal sounds good.

rschu1ze committed:
Update cmake/ccache.cmake

Co-authored-by: Azat Khuzhin <a3at.mail@gmail.com>

rschu1ze committed:
Remove unused file

rschu1ze committed:
Remove performance workaround for old GCC

GCC support these days is experimental. GCCs main use is to keep the
code portable but I don't think it's used for performance tests. Hence
removing a performance workaround added in the GCC 7 days (we are now
using 11, soon: 12).

alexey-milovidov commented:
Interesting, how did we miss `-Wextra-semi`?

alexey-milovidov commented on cmake/ccache.cmake:
```diff
@@ -25,25 +25,27 @@ endif()
+message(STATUS "Using ccache: ${CCACHE_EXECUTABLE} (version ${CCACHE_VERSION})")
+set(LAUNCHER ${CCACHE_EXECUTABLE})
+
+# debian (debhelpers) set SOURCE_DATE_EPOCH environment variable, that is
```
Just in case, we don't use debhelpers anymore.

rschu1ze replied:
Clarified the comment.

alexey-milovidov commented on cmake/warnings.cmake:
```diff
@@ -9,93 +9,46 @@
 endif ()
 
 if (COMPILER_CLANG)
+    add_warning(everything)
```
I'm actually very proud of our usage of `-Weverything`
let's keep the comment.

alexey-milovidov replied:
Added a comment.

alexey-milovidov committed:
Update warnings.cmake

rschu1ze committed:
Small update of comment for SOURCE_DATE_EPOCH workaround

azat commented on cmake/ccache.cmake:
```diff
@@ -1,56 +1,53 @@
+#
+# (*) https://reproducible-builds.org/specs/source-date-epoch/
+if (CCACHE_VERSION VERSION_GREATER_EQUAL "4.0" AND CCACHE_VERSION VERSION_LESS "4.2")
+    message(STATUS "Ignore SOURCE_DATE_EPOCH for ccache 4.1 / 4.2")
```
```suggestion
    message(STATUS "Ignore SOURCE_DATE_EPOCH for ccache 4.0 / 4.1")
```

rschu1ze replied:
Thanks (guess I need to get another coffee  :D).

rschu1ze committed:
Update cmake/ccache.cmake

Co-authored-by: Azat Khuzhin <a3at.mail@gmail.com>

rschu1ze commented:
> Interesting, how did we miss `-Wextra-semi`?

As far as I see, it is enabled by "-wc++98-compat" which was (rightfully) disabled in the WEVERYTHING path. Will check independently if the warning can be enabled using some other check. But in the end it is cosmetics only.
